### PR TITLE
Add ScanVarLen op (opset 27)

### PR DESCRIPTION
**Pull request by gramalingam**
## Summary

Adds `ScanVarLen`, a variation of the `Scan` op that supports variable-length scan outputs, at opset 27 in the ai.onnx standard domain. Based on the design proposed in microsoft/onnxruntime#27761 (which placed it in `com.microsoft`); this PR brings the same op into the standard ONNX domain.

## Motivation

Encodes computations over batches of variable-length / ragged sequences (e.g. PackedAttention's compact representation). Any processing of a ragged batch to produce another ragged batch can be expressed as a scan but requires variable-length scan outputs — which standard Scan cannot express.

## Design

- New op at opset 27, ai.onnx domain. Strictly additive — standard `Scan` v25 is untouched.
- Differences from standard Scan: (1) prepended optional `output_lengths` input (`int64[K]` hint enabling pre-allocation); (2) no `scan_output_directions` attribute (always forward); (3) concat-not-stack semantics — each iteration's body output is concatenated along `scan_output_axes[i]` (the axis already exists in the body output) rather than stacked along a new dim. Generalizes Scan's "size-1 contribution per iteration" to "any size ≥ 0".
- Rank-preserving: output rank equals body output rank.
- Ships as a primitive (no `FunctionBody`), matching standard Scan.
- `output_lengths` mismatch is a runtime error (hard contract).
- **Zero iterations are a hard error.** ScanVarLen requires `sequence_length >= 1`. Shape inference fails when the sequence-axis dim is statically 0; runtimes raise at execution time when it is dynamically 0. Models that need to handle empty sequences should guard the ScanVarLen call with an `If` node. Rationale: supporting zero iterations would either force runtimes to execute the body's shape inference at runtime to fabricate empty outputs with the correct non-concat dimensions, or leave the body's execution status implementation-defined (as standard Scan v25 latently does). Either alternative complicates the runtime contract; requiring an explicit `If`-guard keeps the semantics simple and unambiguous.

## Files changed (high level)

- C++ schema, named shape inference function, doc strings (`onnx/defs/controlflow/{defs,utils}.cc` + `onnx/defs/doc_strings.cc` + `onnx/defs/operator_sets.h`)
- Python reference implementation (`onnx/reference/ops/op_scan_var_len.py`)
- 6 node tests covering basic / output_lengths / reverse / non-default-axes / state / ragged
- Python shape-inference tests (positive + negative, incl. static-zero-iter rejection and symbolic-sequence-length acceptance) + C++ shape-inference gtests (incl. static-zero-iter fails)
- Reference-evaluator end-to-end test + `output_lengths`-mismatch test + zero-iter runtime-error test
- Auto-upgrade test
- Regenerated `docs/Operators.md`, `docs/Changelog.md`, `docs/TestCoverage.md`

## Testing

- `lintrunner -a` clean
- C++ gtests pass (including 5 ScanVarLen tests, one of which exercises the new static-zero-iter shape-inference failure)
- Full pytest sweep green
- Targeted ScanVarLen suites: shape-inf (positive + negative) + ref-eval + auto-upgrade + node-test backend-ref + node-test backend all green

## Spec decisions worth flagging for ONNX operator-SIG

1. `output_lengths` mismatch = runtime error (vs UB). Mirrors PR microsoft/onnxruntime#27761 kernel.
2. **`sequence_length == 0` is a hard error** (shape inference fails when statically 0; runtime raises when dynamically 0). The contrib op in microsoft/onnxruntime#27761 left this case unhandled; we make it explicit so users have an unambiguous contract. Workaround for callers that need empty-sequence support: wrap ScanVarLen in an `If` node that constructs an empty output on the `sequence_length == 0` branch.
3. Differentiability tags match standard Scan (Unknown on variadic state+scan I/O; `NonDifferentiable` on `output_lengths`).
4. Input named `output_lengths` to match PR microsoft/onnxruntime#27761 (clarified in doc as per-output-total, not per-iteration). Some reviewers suggested `scan_output_total_sizes` — feedback welcome.

## Reference

Design source: microsoft/onnxruntime#27761

### CODE REVIEW

File: onnx/defs/doc_strings.cc [hunk 1]
```diff
@@ -779,6 +779,123 @@ values are computed in the outer graph, they need to be passed in as extra state
...
+    }
+
+    // If output_lengths was provided, the concatenation-axis size of each
+    // scan_out_i is required to equal output_lengths[i]; otherwise a runtime
+    // error is raised.
+
+    return st_1, ..., st_n, scan_out_1, ..., scan_out_k;
+
+If `sequence_length` is 0, each scan output has a 0-sized dimension at
+`scan_output_axes[i]`. Conforming implementations may or may not execute the body
```
Comment (gramalingam) [inline onnx/defs/doc_strings.cc]: Unclear how an implementation can execute the body (since there are no valid inputs to execute it). Better to omit this sentence.

File: onnx/defs/doc_strings.cc [hunk 2]
```diff
@@ -779,6 +779,123 @@ values are computed in the outer graph, they need to be passed in as extra state
...
+subgraph in this case (implementation-defined); the output shape is determined by the
+body's output shape with the concat-axis dimension set to 0. Shape inference still runs
```
Comment (gramalingam) [inline onnx/defs/doc_strings.cc]: The output shapes when sequence_length is 0 seems tricky. The specification above makes sense, but this seems to be the one exception where the shapes declared in the graph has a runtime semantic significance (beyond just correctness checking). I wonder if it is better to adopt this or disallow the sequence_length is 0 case?

**Review by cbourjau — CHANGES_REQUESTED** [review]
Please use our [RFC template](https://github.com/onnx/onnx/blob/main/docs/proposals/0000-template.md) when proposing a new operator. It makes discussion and review of the design much more efficient and much less error prone.

**Comment by gramalingam** [comment]
> Please use our [RFC template](https://github.com/onnx/onnx/blob/main/docs/proposals/0000-template.md) when proposing a new operator. It makes discussion and review of the design much more efficient and much less error prone.

Not sure I agree with the conclusion that it makes the discussion/review much more efficient. I think the PR description above is a compact version of a design proposal, with details in the code. I am happy to discuss it in the operator SIG meeting to answer any questions. Personally, I would be glad to see some review comments here, and, sadly, there are none.

**Comment by cbourjau** [comment]
> Not sure I agree with the conclusion that it makes the discussion/review much more efficient. I think the PR description above is a compact version of a design proposal, with details in the code. I am happy to discuss it in the operator SIG meeting to answer any questions. Personally, I would be glad to see some review comments here, and, sadly, there are none.

I'd really like to give feedback, but if I now start to quote-comment from the PR description, then everything is just going to diverge immediately. Discussing the semantics along with ~2500 LoCs is also inevitably going to spread attention to detail thin. Discussing operators purely in meetings is problematic since it doesn't leave an obvious place to put that information for future reference and details are quickly lost when only discussed verbally. In comparison having a PR + discussion on a single RFC-md file is well scoped and, IMHO, more efficient. The template also points out common questions that have been frequently overlooked in the past and that are not sufficiently addressed in the description of this PR, either, IMHO. For instance, I'm wondering what the alternatives and prior art are. Is there something similar in PyTorch, for instance? What other (viable) alternatives were considered and why was this implementation chosen over them? Why is `SequenceMap` not sufficient? Again, I'd love to give proper feedback, but I really don't think that GitHub PR descriptions are meant to be the main focus of a GitHub PR discussion ;)

**Comment by gramalingam** [comment]
> Discussing the semantics along with ~2500 LoCs is also inevitably going to spread attention to detail thin. Discussing operators purely in meetings is problematic since it doesn't leave an obvious place to put that information for future reference and details are quickly lost when only discussed verbally. In comparison having a PR + discussion on a single RFC-md file is well scoped and, IMHO, more efficient

My concerns are as follows. A meeting helps make progress. Asynchronous discussions happening via PRs can sometimes take a long time and get stuck.

As one example: https://github.com/onnx/onnx/pull/7661 was a PR ... and, in the spirit of RFC proposal recently suggested, I asked copilots to create a RFC, which is https://github.com/onnx/onnx/pull/7692 ... created in Feb, it has got stuck and not progressed. I think an in-meeting discussion may be the best way to make progress on that.

I could ask copilots to create a RFC for current PR too. But based on several recent experience, it is not going to be perfect. It will still require a substantial human effort to make it acceptable to everyone. 

My concern is that the RFC process, though well intentioned, is going to make PR progress much slower. Right now, this is going to pretty much mean it won't make it to the release cut planned for June 1st. I am hoping the meeting on thursday will give us a chance of making a quick call on how to move forward with this.

**Comment by gramalingam** [comment]
> Why is `SequenceMap` not sufficient?

For example, this is much easier to discuss in a meeting

**Review by justinchuby — COMMENTED** [review]
# Copilot

## Overall

Solid PR. The design is sound, the additive-only approach (new op, no changes to `Scan` v25) keeps the blast radius small, and the test coverage looks broad (6 node tests + Python shape-inf positive/negative + C++ shape-inf gtests + ref-eval end-to-end + auto-upgrade + the regenerated docs). Comments are mostly nits / spec-level pushback worth raising for SIG.

## Spec / contract questions

1. **Rename `output_lengths`.** As flagged in the PR description, the name reads as "per-iteration" until you read the docstring. `scan_output_total_sizes` (or `scan_output_concat_sizes`) is unambiguous and matches the shape contract precisely. I'd take the small rename hit here rather than carrying ambiguity into the standard domain forever.

2. **Mismatch-detection timing.** The doc says "It is a runtime error if `output_lengths[i]` does not match the actual concatenation size". Implementations that pre-allocate from the hint will know mid-loop if the body overshoots (overflow), but undershoots only become observable after the final iteration. Worth saying explicitly whether implementations *must* raise as early as they can detect it, or whether deferring to post-loop is conforming. Either is defensible, but the spec should pick one.

3. **Zero-iteration as a hard error.** The reasoning (don't force the runtime to fabricate empty-output shapes via body shape inference) is convincing. Worth confirming microsoft/onnxruntime#27761's kernel agrees — the PR description says the contrib op left this case unhandled, and ONNX is now tightening it. If the contrib op is already shipping in models, that's a soft incompatibility we should call out.

4. **Differentiability.** PR description claims "Unknown on variadic state+scan I/O" — but the schema doesn't tag the variadic input/output at all. That's the same as Scan v25, so it's consistent, but the claim could mention that `Unknown` here means "not annotated; matches Scan v25", not "explicitly tagged Unknown".

## Code / test gaps

5. **No node test for `output_lengths` mismatch.** The mismatch case is only exercised in `test/reference_evaluator_test.py::test_scan_var_len_output_lengths_mismatch`, which is Python-only. Backends consuming the `node/test_data/` test cases have no oracle for this contract. Suggest adding a `scan_var_len_output_lengths_mismatch` node test that's expected to raise (or at minimum mark this as a documented spec test that backends can opt to validate).

6. **Ragged test could include a zero-length contribution.** `test_scan_var_len_ragged` uses `lengths=[[1],[2],[3]]` — all non-zero. The doc explicitly allows per-iteration contributions of size 0 along the concat axis. An extra row with `length=0` would exercise that path and lock the contract for backends. (You'd still need `sum >= 1` to satisfy the sequence_length ≥ 1 rule, just have one iteration contribute zero.)

7. **`op_scan_var_len.py:243` accepts any integer dtype**, but the schema's type constraint is `tensor(int64)` only. Tightening to `arr.dtype == np.int64` keeps the reference impl honest with the spec — the checker will reject mismatched dtypes upstream anyway, so this only fires for hand-constructed feeds, but it's a useful guard.

## Minor

8. **`utils.cc:489-521`** — `first_zero_scan_input` tracks the first scan input with a statically-0 sequence dim, and the error message names it. But `mergeInDimensionInfo` would fail before line 510 if scan inputs disagree (one has static 0 and another has static N > 0), so by the time we're in this branch, *every* scan input with a known sequence dim is 0. The "scan input X" attribution is technically correct (it's the first one encountered) but reads as if there might be some non-zero ones too. Consider rewording to "(at least) scan input X" or simply emit the simpler error message used in the `else` branch and drop the `first_zero_scan_input` bookkeeping.

9. **`utils.cc:531`** — passing `nullptr` for each subgraph input data is correct for scan inputs (they're sliced per iteration), but the state variables *could* be const-propagated for shape inference. Standard `Scan` does the same thing, so this is consistent — just noting in case a follow-up tightens it.

10. **`op_scan_var_len.py:115-120`** — error message says "expected at least N (one per loop state variable)" but the body must emit N + K outputs, not "at least N". Suggest "expected exactly N + K (N=loop_state + K=scan_outputs)". The check itself is fine; just the message is loose.

11. **`scan_var_len.py:24`** — `_OPSET_IMPORTS = [onnx.helper.make_opsetid("", 27)]` works through the "unreleased opset" branch in `node/__init__.py:228-249`. Just confirming this is intentional (it is) and that the next `VERSION_TABLE` row bumps to opset 27 at release time.

## Things that look good

- Strict additivity (no churn in Scan v25).
- Rank-preserving semantics — much easier to teach than Scan's stack-then-collapse.
- Shape inference both fails loudly on static-0 sequence and accepts symbolic sequence (covered by tests on both sides).
- Doc explicitly distinguishes from Scan with the equivalence pseudo-code — that's exactly the right level of formality for a new control-flow primitive.

Nice work.

**Review by justinchuby — COMMENTED** [review]
I reviewed the spec. LGTM

**Review by justinchuby — APPROVED** [review]
I reviewed the spec manually. LGTM

File: onnx/reference/ops/op_scan_var_len.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    """
+    normalized = axis + rank if axis < 0 else axis
+    if not 0 <= normalized < rank:
+        raise ValueError(
+            f"ScanVarLen: {attr_name}[{index}]={axis} is out of range for rank {rank}."
+        )
+    return normalized
+
+
+class ScanVarLen(OpRun):
```
Comment (justinchuby) [inline onnx/reference/ops/op_scan_var_len.py:141]: Is `need_context` needed? Similar to https://github.com/onnx/onnx/pull/7964

**Review by cbourjau — CHANGES_REQUESTED** [review]
> > Discussing the semantics along with ~2500 LoCs is also inevitably going to spread attention to detail thin. Discussing operators purely in meetings is problematic since it doesn't leave an obvious place to put that information for future reference and details are quickly lost when only discussed verbally. In comparison having a PR + discussion on a single RFC-md file is well scoped and, IMHO, more efficient
> 
> My concerns are as follows. A meeting helps make progress. Asynchronous discussions happening via PRs can sometimes take a long time and get stuck.
> 

My argument is that there should be both: A solid written RFC for future reference and to serve as basis for frutiteful discussions in meetings.

> As one example: #7661 was a PR ... and, in the spirit of RFC proposal recently suggested, I asked copilots to create a RFC, which is #7692 ... created in Feb, it has got stuck and not progressed. I think an in-meeting discussion may be the best way to make progress on that.
> 

The proposal is stuck for a reason: as proposed, I can't see this proposal actually solving the problem it is trying to solve. Rather than stabilizing something that has to be supported for all time to come, we need to find a better approach. It is obviously not the intention to have to parallel PRs, but in general I'd say the RFC works as intended, to be honest. I think this is something that needs to be improved in the operator-sig. If people don't review the proposals in writing, then that also means that they are likely not in a position to give informed feedback in the meetings.

> I could ask copilots to create a RFC for current PR too. But based on several recent experience, it is not going to be perfect. It will still require a substantial human effort to make it acceptable to everyone.
> 

New operators have to be supported for all times to come, and we expect downstream runtimes to implement and support them basically indefinitely, too. There is a huge cost involved in stabilizing operators that turn out not to hit the mark.

> My concern is that the RFC process, though well intentioned, is going to make PR progress much slower. Right now, this is going to pretty much mean it won't make it to the release cut planned for June 1st. I am hoping the meeting on thursday will give us a chance of making a quick call on how to move forward with this.

We increased the release cadence to avoid exactly this kind of issue. The addition of new operators will be more efficient and faster if it follows the proposal path rather than doing it freestyle.

>>    Why is SequenceMap not sufficient?

> For example, this is much easier to discuss in a meeting

Thoroughly discussing alternatives (especially ones already available in the standard) seems to be one of the most important points when proposing an operator. I don't understand why we can't put that into writing to make sure that we get this right.

File: onnx/defs/doc_strings.cc [hunk 4]
```diff
@@ -779,6 +779,132 @@ values are computed in the outer graph, they need to be passed in as extra state
...
+        scan_out_k = Concat<axis=out_axis_k>(scan_out_k, so_k);
+    }
+
+    // If output_lengths was provided, the concatenation-axis size of each
+    // scan_out_i is required to equal output_lengths[i]; otherwise a runtime
+    // error is raised.
+
+    return st_1, ..., st_n, scan_out_1, ..., scan_out_k;
+
+ScanVarLen requires `sequence_length >= 1` (i.e. at least one iteration). It is an
```
Comment (xadupre) [inline onnx/defs/doc_strings.cc]: Do we accept an iteration which produces outputs with empty shapes?

Comment (gramalingam) [reply onnx/defs/doc_strings.cc]: yes, that's ok

**Comment by gramalingam** [comment]
Converting to draft, given ongoing discussion
